qa/integration-tests/src/test/java/io/camunda/it/client/ProcessQueryTest.java

```diff
@@ -0,0 +1,567 @@
+    final List<String> sortedDesc = allStates.stream().sorted(Comparator.reverseOrder()).toList();
+
+    // then
+    assertThat(resultAsc.items().stream().map(ProcessInstance::getState).toList())
+        .containsExactlyElementsOf(sortedAsc);
+    assertThat(resultDesc.items().stream().map(ProcessInstance::getState).toList())
+        .containsExactlyElementsOf(sortedDesc);
+  }
+
+  @Test
+  public void shouldValidatePagination() {
+    final var result =
+        zeebeClient
+            .newProcessInstanceQuery()
+            .filter(f -> matchAllProcessInstancesFilter(f))
+            .page(p -> p.limit(2))
+            .send()
+            .join();
+    assertThat(result.items().size()).isEqualTo(2);
+    final var key = result.items().getFirst().getKey();
+    // apply searchAfter
+    final var resultAfter =
+        zeebeClient
+            .newProcessInstanceQuery()
+            .filter(f -> matchAllProcessInstancesFilter(f))
+            .page(p -> p.searchAfter(Collections.singletonList(key)))
+            .send()
+            .join();
+
+    assertThat(resultAfter.items().size()).isEqualTo(5);
+    final var keyAfter = resultAfter.items().getFirst().getKey();
+    // apply searchBefore
+    final var resultBefore =
+        zeebeClient
+            .newProcessInstanceQuery()
+            .filter(f -> matchAllProcessInstancesFilter(f))
+            .page(p -> p.searchBefore(Collections.singletonList(keyAfter)))
+            .send()
+            .join();
+    assertThat(result.items().size()).isEqualTo(2);
+    assertThat(resultBefore.items().getFirst().getKey()).isEqualTo(key);
+  }
+
+  private static void waitForProcessesToBeDeployed() {
+    // Waiting here should be done with zeebeClient.newProcessQuery() but it is not implemented yet
+    final long waitDuration = 15;
+    Awaitility.await("should deploy process definitions and import in Operate")
+        .pollDelay(Duration.ofSeconds(waitDuration))
+        .timeout(Duration.ofSeconds(waitDuration + 10))
+        .until(() -> true);
+  }
+
+  private static void waitForProcessInstancesToStart() {
+    Awaitility.await("should start process instances and import in Operate")
+        .atMost(Duration.ofSeconds(60))
+        .ignoreExceptions() // Ignore exceptions and continue retrying
+        .untilAsserted(
+            () -> {
+              final var result =
+                  zeebeClient
+                      .newProcessInstanceQuery()
+                      .filter(f -> matchAllProcessInstancesFilter(f))
+                      .send()
+                      .join();
+              assertThat(result.items().size()).isEqualTo(6);
+            });
+  }
+
+  private static void waitForProcessInstancesToExecute() {
+    final long waitDuration = 15;
+    Awaitility.await("should execute process instances and import in Operate")
+        .pollDelay(Duration.ofSeconds(waitDuration))
+        .timeout(Duration.ofSeconds(waitDuration + 10))
+        .until(() -> true);
```

Comment:
If I understand it, this just waits for 15 seconds. Can we just replace it with `Thread.sleep(15000);` ?
is there any other conditions we can check to reduce this wait time ?

Reply:
Unfortunately here there is no easy way to check that the execution completed for all the process instances involved. Some processes have incidents, so they are not actually completed in their expected end state. I know the expected end state, but as I also mentioned in Slack, this state should be checked with an external tool, like ES client. Currently we do not have this option.